### PR TITLE
Add support for tags for clients that allow them (e.g. datadog)

### PR DESCRIPTION
**Pull request by timtrinidad**
DatadogStatsd allows tags to be sent along with regular statsd metrics. Add support for tags and append them to the message.

### CODE REVIEW

File: lib/Client.php [hunk 1]
```diff
@@ -271,6 +284,46 @@ private function send($key, $value, $type, $sampleRate)
...
+    private function generateTagsMessage($tags)
+    {
+      // Nothing to return if no tags
+      if (empty($tags)) {
+        return '';
+      }
+
+      // Normalize string inputs into an array
+      if (!is_array($tags)) {
+        $tags = [$tags];
```
Comment (delphyne) [inline lib/Client.php:299]: do you really want to do this?  you'll end up with a tag whose name is "0" and whose value is the bogus thing passed in.

Comment (delphyne) [reply lib/Client.php:299]: looking at the test, it seems that it ends up being named the same thing as the value... can't honestly say i understand why, but you can disregard my comment.

**Review by delphyne — APPROVED** [review]
lgtm

File: lib/Client.php (hunk 1 above)
Comment (timtrinidad) [reply lib/Client.php:299]: Based on [this link](https://help.datadoghq.com/hc/en-us/articles/206441345-Send-metrics-and-events-using-dogstatsd-and-the-shell), they accept valueless tags. I also modeled after the [tag handing logic](https://github.com/DataDog/php-datadogstatsd/blob/master/libraries/datadogstatsd.php#L150) in the datadog/php-datadogstatsd repo.

File: lib/Client.php [hunk 2]
```diff
@@ -271,6 +284,46 @@ private function send($key, $value, $type, $sampleRate)
     }
 
     /**
+     * @param $tags
```
Comment (jroman00) [inline lib/Client.php]: Missing type: `array|string`

File: lib/Client.php [hunk 3]
```diff
@@ -271,6 +284,46 @@ private function send($key, $value, $type, $sampleRate)
...
+     * @return string
+     */
+    private function generateTagsMessage($tags)
+    {
+      // Nothing to return if no tags
+      if (empty($tags)) {
```
Comment (jroman00) [inline lib/Client.php]: Optionally skip the `empty` check and just do `if (!$tags)`.

File: lib/Client.php [hunk 4]
```diff
@@ -271,6 +284,46 @@ private function send($key, $value, $type, $sampleRate)
...
+        }
+      }
+
+      return '|#' . implode(',', $tagsMessages);
+    }
+
+    /**
+     * Clean up a tag value
+     */
+    private function escapeTag($tag)
```
Comment (jroman00) [inline lib/Client.php]: Missing `@param` and `@return`. I know it's a private method, but why not?

File: tests/unit/ClientTest.php [hunk 5]
```diff
@@ -226,4 +226,39 @@ public function testSet()
...
+     */
+    public function tagsProvider()
+    {
+        return [
+          // single string tag
+          ['one', 'one'],
+
+          // basic arrays
+          [['one'], 'one'],
+          [['one', 'two'], 'one,two'],
```
Comment (jroman00) [inline tests/unit/ClientTest.php]: What's expected when values are duplicated? Might not be a valid use case. Worth testing? e.g. `[['one', 'one'], 'one,one'],`

File: lib/Client.php (hunk 4 above)
Comment (jroman00) [inline lib/Client.php]: You're not "escaping" here as much as you are "cleaning" or "sanitizing".

**Review by jroman00 — CHANGES_REQUESTED** [review]

**Review by jroman00 — APPROVED** [review]
:monkey:

**Pull request merged**